extras/module-node/duk_module_node.c

```diff
@@ -0,0 +1,248 @@
+	duk_push_string(ctx, "require");
+	duk_def_prop(ctx, -3, DUK_DEFPROP_HAVE_VALUE);
+
+	/* require.cache */
+	duk_push_global_stash(ctx);
+	duk_get_prop_string(ctx, -1, "\xffModuleCache");
+	duk_put_prop_string(ctx, -3, "cache");
+	duk_pop(ctx);
+}
+
+static void duk__push_module_object(duk_context *ctx, const char *id) {
+	duk_push_object(ctx);
+
+	/* module.exports = {} */
+	duk_push_object(ctx);
+	duk_put_prop_string(ctx, -2, "exports");
+
+	/* module.filename = id */
+	duk_push_string(ctx, id);
+	duk_put_prop_string(ctx, -2, "filename");
+
+	/* module.id = id */
+	duk_push_string(ctx, id);
+	duk_put_prop_string(ctx, -2, "id");
+
+	/* module.loaded = false */
+	duk_push_false(ctx);
+	duk_put_prop_string(ctx, -2, "loaded");
+
+	/* module.require */
+	duk__push_require_function(ctx, id);
+	duk_put_prop_string(ctx, -2, "require");
+}
+
+static duk_int_t duk__eval_module_helper(duk_context *ctx) {
+	/*
+	 *  Stack: ... module source --> ... module (retval == 0)
+	 *         ... module source --> ... err    (retval != 0)
+	 */
+
+	/*
+	 * Synthesize a function to wrap the module code.  This is the easiest way
+	 * to implement CommonJS semantics and matches the behavior of Node.js.
+	 * The wrapper has no line breaks so it won't interfere with line
+	 * numbering.
+	 */
+	duk_push_string(ctx, "(function main(exports, require, module, __filename) { ");
+	duk_dup(ctx, -2);  /* source */
+	duk_push_string(ctx, " })");
+	duk_concat(ctx, 3);
+
+	/* module source func_src */
+
+	duk_remove(ctx, -2);
+
+	/* module func_str */
+
+	duk_get_prop_string(ctx, -2, "filename");
+	if (duk_pcompile(ctx, DUK_COMPILE_EVAL) != DUK_EXEC_SUCCESS) {
+		duk_remove(ctx, -2);
+		goto fail;  /* leave error on stack */
+	}
+	duk_call(ctx, 0);
+
+	/* module func */
+
+	/* Call the function wrapper */
+	duk_get_prop_string(ctx, -2, "exports");   /* exports */
+	duk_get_prop_string(ctx, -3, "require");   /* require */
+	duk_dup(ctx, -4);                          /* module */
+	duk_get_prop_string(ctx, -5, "filename");  /* __filename */
+	if (duk_pcall(ctx, 4) != DUK_EXEC_SUCCESS) {
+		duk_remove(ctx, -2);
+		goto fail;  /* leave error on stack */
+	}
+	duk_pop(ctx);
+
+	/* module */
+
+	duk_push_true(ctx);
+	duk_put_prop_string(ctx, -2, "loaded");
+
+	return DUK_EXEC_SUCCESS;
+
+fail:
+	/* err */
+
+	return DUK_EXEC_ERROR;
+}
+
+void duk_module_node_init(duk_context *ctx, duk_resolve_module_function resolve_cb, duk_load_module_function load_cb) {
+	resolve_module_cb = resolve_cb;
+	load_module_cb = load_cb;
+
+	/* create an object to store cached modules */
+	duk_push_global_stash(ctx);
+	duk_push_object(ctx);
+	duk_put_prop_string(ctx, -2, "\xffModuleCache");
+	duk_pop(ctx);
+
+	/* push the default `require` builtin */
+	duk_push_global_object(ctx);
+	duk_push_string(ctx, "require");
+	duk__push_require_function(ctx, "");
+
+	/* global `require` is not enumerable */
+	duk_def_prop(ctx, -3, DUK_DEFPROP_HAVE_VALUE
+		| DUK_DEFPROP_SET_WRITABLE
+		| DUK_DEFPROP_SET_CONFIGURABLE);
```

Comment:
The indent style makes the flags a bit difficult to read. The typical style for this would be to tab indent for the base level (`duk_def_prop()`) and then space indent to align the rest:

    duk_def_prop(ctx, -2, DUK_DEFPROP_HAVE_VALUE |
                          DUK_DEFPROP_SET_WRITABLE |
                          DUK_DEFPROP_SET_CONFIGURABLE);